internal/controller/dnsrecord_healthchecks.go

```diff
@@ -0,0 +1,191 @@
+	}
+
+	result := r.reconcileHealthCheckStatus(results, dnsRecord)
+	return result
+}
+
+func (r *DNSRecordReconciler) getProbeStatus(address string, dnsRecord *v1alpha1.DNSRecord) *v1alpha1.HealthCheckStatusProbe {
+	if dnsRecord.Status.HealthCheck == nil || dnsRecord.Status.HealthCheck.Probes == nil {
+		return nil
+	}
+	for _, probeStatus := range dnsRecord.Status.HealthCheck.Probes {
+		if probeStatus.IPAddress == address {
+			return &probeStatus
+		}
+	}
+
+	return nil
+}
+
+func (r *DNSRecordReconciler) reconcileHealthCheckStatus(results []provider.HealthCheckResult, dnsRecord *v1alpha1.DNSRecord) error {
+	var previousCondition metav1.Condition
+	probesCondition := metav1.Condition{
+		Reason: "AllProbesSynced",
+		Type:   "healthProbesSynced",
+	}
+
+	var allSynced = metav1.ConditionTrue
+
+	if dnsRecord.Status.HealthCheck == nil {
+		dnsRecord.Status.HealthCheck = &v1alpha1.HealthCheckStatus{
+			Conditions: []metav1.Condition{},
+			Probes:     []v1alpha1.HealthCheckStatusProbe{},
+		}
+	}
+
+	for _, c := range dnsRecord.Status.HealthCheck.Conditions {
+		if c.Type == "healthProbesSynced" {
+			previousCondition = c
+			probesCondition = previousCondition
+		}
+	}
+
+	dnsRecord.Status.HealthCheck.Probes = []v1alpha1.HealthCheckStatusProbe{}
+
+	for _, result := range results {
+		status := true
+		if result.Result == provider.HealthCheckFailed {
+			status = false
+			allSynced = metav1.ConditionFalse
+		}
+
+		dnsRecord.Status.HealthCheck.Probes = append(dnsRecord.Status.HealthCheck.Probes, v1alpha1.HealthCheckStatusProbe{
+			ID:        result.ID,
+			IPAddress: result.IPAddress,
+			Host:      result.Host,
+			Synced:    status,
+		})
+	}
+
+	probesCondition.ObservedGeneration = dnsRecord.Generation
+	probesCondition.Status = allSynced
+
+	if allSynced == metav1.ConditionTrue {
+		probesCondition.Message = fmt.Sprintf("all %v probes synced successfully", len(dnsRecord.Status.HealthCheck.Probes))
+		probesCondition.Reason = "AllProbesSynced"
+	} else {
+		probesCondition.Reason = "UnsyncedProbes"
+		probesCondition.Message = "some probes have not yet successfully synced to the DNS Provider"
+	}
+
+	//probe condition changed? - update transition time
+	if !reflect.DeepEqual(previousCondition, probesCondition) {
+		probesCondition.LastTransitionTime = metav1.Now()
+	}
+
+	dnsRecord.Status.HealthCheck.Conditions = []metav1.Condition{probesCondition}
+
+	return nil
+}
+
+func getHealthChecksConfig(dnsRecord *v1alpha1.DNSRecord) *healthChecksConfig {
+	if dnsRecord.Spec.HealthCheck == nil || dnsRecord.DeletionTimestamp != nil {
+		return nil
+	}
+
+	return &healthChecksConfig{
+		Endpoint:         dnsRecord.Spec.HealthCheck.Endpoint,
+		Port:             valueAs(toInt64, dnsRecord.Spec.HealthCheck.Port),
+		FailureThreshold: valueAs(toInt64, dnsRecord.Spec.HealthCheck.FailureThreshold),
+		Protocol:         (*provider.HealthCheckProtocol)(dnsRecord.Spec.HealthCheck.Protocol),
+	}
+}
+
+func valueAs[T, R any](f func(T) R, original *T) *R {
```

Comment:
seems a bit overkill ?

Reply:
I think this is just making it a little more legible and compact to convert stuff to pointers, does it need changing here?